### PR TITLE
PHPLIB-617 Continuous Matrix Testing for 4.0 era drivers

### PR DESCRIPTION
**Pull request by alcaeus**
PHPLIB-617

Since this version is not tested on evergreen yet, this includes the minimal scripts required to test on evergreen without adding an evergreen configuration.

### CODE REVIEW

File: .evergreen/run-tests.sh
Comment (alcaeus) [inline .evergreen/run-tests.sh:23]: This is necessary since the drivers-matrix-testing project does not allow to call additional scripts. I'll propose changes to call install-dependencies as that's a standard script from drivers-evergreen-tools that all drivers should be using.

Allowing projects to define their own environment variables also removes the need for hardcoding an extension version here.

File: .evergreen/run-tests.sh
Comment (alcaeus) [inline .evergreen/run-tests.sh:16]: The drivers-matrix-testing project does not allow to set additional environment variables. I'll work on adding support for that next, until then we need this fallback.

**Review by alcaeus — COMMENTED** [review]
@jmikola a couple of notes on the scripts since they introduce changes from the ones we use in master.

File: .evergreen/install-dependencies.sh
Comment (jmikola) [inline .evergreen/install-dependencies.sh:40]: Would it make more sense to _append_ our `php.ini` to the existing one, or is there nothing of value in the original `php.ini`?

I looked at https://github.com/10gen/mongo-php-toolchain but I don't see any reference to `php.ini` at all. I dunno if that's something we might want to change down the line, as I happen to use the following for my local PHP builds:

```
date.timezone = "America/New_York"
display_errors = On
display_startup_errors = On
error_reporting = -1
log_errors = On
memory_limit = -1
report_memleaks = On
report_zend_debug = 0
short_open_tag = Off
```

Obviously `date.timezone` isn't relevant, but I think the other settings could be helpful to ensure PHP doesn't suppress memory leaks and such by default.

File: .evergreen/run-tests.sh
Comment (jmikola) [inline .evergreen/run-tests.sh:22]: Ah, we lucked out with this being 1.5+, as we only standardized on the `configure` options in 1.5.0: https://github.com/mongodb/mongo-php-driver/commit/da72946e654f7a5afda189795a05a03726c1877f

File: .evergreen/install-dependencies.sh
Comment (jmikola) [inline .evergreen/install-dependencies.sh:95]: I assume htis was lifted from https://getcomposer.org/download/. Perhaps it makes more sense to follow the instructions in https://getcomposer.org/doc/faqs/how-to-install-composer-programmatically.md so we can check the signature? Looks like it'd be easy to package into another shell script.

File: .evergreen/run-tests.sh
Comment (jmikola) [inline .evergreen/run-tests.sh:19]: I assume `PLATFORM` is actually used for some other purpose. Is there a ticket to introduce another environment variable (e.g. `MATRIX_TESTING`) specifically so scripts can determine if they're running for matrix testing?

File: .evergreen/run-tests.sh
Comment (jmikola) [inline .evergreen/run-tests.sh:28]: Should we rephrase this? I imagine this might print: "Running noauth tests over nossl", which seems odd.

What are `AUTH` and `SSL` when they are set (and not defaulting to `noauth` and `nossl`, respectively)?

Perhaps something like:

* `Running tests with AUTH=$AUTH, SSL=$SSL, connecting to $MONGO_URI`

Or if `$AUTH` (and `$SSL` would only ever be) "auth" or "noauth", then perhaps:

* `Running tests with $AUTH and $SSL, connecting to: $MONGODB_URI`

File: .evergreen/install-dependencies.sh
Comment (jmikola) [inline .evergreen/install-dependencies.sh:37]: I know PECL tries to append to `php.ini` after it runs. It looks like that only if `php_ini` is specified in the PEAR/PECL configuration.

If you want to add `pecl config-set php_ini ${PHP_PATH}/lib/php.ini` before invoking `pecl install`, perhaps that would let you do without the custom `php.ini` file in this PR. Although if `${PHP_PATH}/lib/php.ini` doesn't already exist (per my next comment), you'd need to `touch` first.

Maybe that's not worth the trouble since you'd also have to deal with file permissions and/or run `pecl` with `sudo`. Feel free to ignore.

File: .evergreen/run-tests.sh
Comment (jmikola) [inline .evergreen/run-tests.sh:39]: I forget when you introduced `simple-phpunit`, but if any of these branches uses that would we want to add an invocation of `vendor/bin/simple-phpunit --version` (nothing that actually runs tests) to `install_dependencies.sh`? IIRC, it installs PHPUnit lazily on its first run. I don't think we'd want that to happen during `run_tests.sh`, since it technically is a dependency.

File: phpunit.evergreen.xml
Comment (jmikola) [inline phpunit.evergreen.xml:18]: Oh, this may relate to my earlier comment about customizing `php.ini`. Are there other options we might want to include here to ensure things like memory leaks get reported? Alternatively, would it make more sense to just move this into the PR's `php.ini` file that you apply to the entire PHP runtime?

File: phpunit.evergreen.xml
Comment (jmikola) [inline phpunit.evergreen.xml:19]: Should this be omitted so we pick up the environment variable from `run_tests.sh`, or is this only a fallback?

File: .evergreen/install-dependencies.sh
Comment (alcaeus) [reply .evergreen/install-dependencies.sh:40]: I haven't focused on improving the process just yet, especially since testing in these branches is difficult as they are not tested on evergreen anyways. We could do the following:
* Move a sensible default configuration to mongo-php-toolchain to have it in place by default, and
* set `php_ini` in the pecl configuration so that `pecl install` enables the extension.

With this, we can remove this workaround, first in master to confirm that everything is working correctly, followed by a backport to these branches. WDYT?

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh:22]: There's a fine line between careful planning and pure dumb luck 😅

File: .evergreen/install-dependencies.sh
Comment (alcaeus) [reply .evergreen/install-dependencies.sh:95]: Ah, very good point. I'll add the script as suggested to the docs here, but this is something we can consider moving to the toolchain (only the script, not the actual composer install)

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh:19]: Yes, this is an ugly monkey patch. `PLATFORM` contains the platform being tested on and is only set by the drivers-matrix-testing pipeline. A more elaborate solution removes the need to call this by being able to define environment variables for driver entries and invoking install-dependencies.sh automatically.

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh:28]: I've noticed the wording before but it has never bugged me to the point where I wanted to change it. Due to evergreen logging everything, this log line is preceded by the actual env variables a few lines above: https://evergreen.mongodb.com/task_log_raw/mongo_php_library_test_php_versions__php_versions~7.0_edge_versions~latest_stable_driver_versions~latest_stable_os_php7~debian92_test_test_replica_set_2d6fce1d04aac91fdc9d5245d5be738555921842_21_02_09_09_52_54/0?type=T#L2266, so we might as well remove this. I'll defer to your preference on this, I typically skip over this line anyways.

File: .evergreen/install-dependencies.sh
Comment (alcaeus) [reply .evergreen/install-dependencies.sh:37]: Also covered by PHPC-1772, and I'd test this in master and then back port.

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh:39]: The v1.7-matrix-testing branch switches to simple-phpunit, the previous branches still use phpunit directly. I'll move installing simple-phpunit to the install-dependencies.sh script so this is not part of running tests 👍

File: phpunit.evergreen.xml
Comment (alcaeus) [reply phpunit.evergreen.xml:18]: This was copied over from phpunit.xml.dist, but I believe we can remove this once we've dealt with php.ini in our toolchain.

File: phpunit.evergreen.xml
Comment (alcaeus) [reply phpunit.evergreen.xml:19]: This is a default so that we don't pass an empty uri if not specified externally. From the [PHPUnit manual on `php.env`](https://phpunit.readthedocs.io/en/9.5/configuration.html#the-env-element):
> By default, environment variables are not overwritten if they exist already. To force overwriting existing variables, use the `force` attribute.

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh]: After some more playing around, I was able to remove the check on PLATFORM and the hardcoded PHP_VERSION. With recent changes in drivers-matrix-testing, custom variables are passed through, so I can leverage the new `PHP_VERSION` and `PHP_IS_MATRIX_TESTING` variables. Since they can only be set per driver (instead of per driverEra, e.g. "4.4"), I can't use the same method for defining the extension version to install, so that's still hardcoded in run-tests.sh.

File: .evergreen/run-tests.sh
Comment (alcaeus) [inline .evergreen/run-tests.sh:16]: This variable will be introduced with https://github.com/mongodb/drivers-matrix-testing/pull/7

File: .evergreen/run-tests.sh
Comment (jmikola) [inline .evergreen/run-tests.sh:22]: Won't this clobber the definitions in https://github.com/mongodb/drivers-matrix-testing/pull/6?

Comment (alcaeus) [reply .evergreen/run-tests.sh:22]: Ah, good point. We probably shouldn't use `export`, but this should be fine without. Alternatively, I can change this from `DRIVER_VERSION` to `EXTENSION_VERSION` (which would be more appropriate anyways). I'll just have to remember to update the config file as well once this gets merged up into a branch with evergreen config present.

File: .evergreen/install-dependencies.sh
Comment (jmikola) [reply .evergreen/install-dependencies.sh:40]: Fine to punt on this. I left a comment on the issue referring back to this thread for context.

File: .evergreen/run-tests.sh
Comment (jmikola) [reply .evergreen/run-tests.sh]: Should this PR wait on https://github.com/mongodb/drivers-matrix-testing/pull/7 to rename the environment variable to `IS_MATRIX_TESTING`?

File: .evergreen/run-tests.sh
Comment (jmikola) [reply .evergreen/run-tests.sh:39]: If the actual environment variables are dumped above, I think we can definitely remove this.

File: .evergreen/run-tests.sh
Comment (jmikola) [reply .evergreen/run-tests.sh:22]: I assume this is just about relaying information to `install-dependencies.sh`, which is called a few lines later. In that case, `EXTENSION_VERSION` seems fine. Using distinct environment variables feels cleaner and could avoid a WTF down the line.

As for merging up to another branch, perhaps you want to lump this todo item in with the simple-phpunit item in https://github.com/mongodb/mongo-php-library/pull/810/files#r574301076?

File: .evergreen/run-tests.sh
Comment (jmikola) [reply .evergreen/run-tests.sh:50]: I'll leave this comment thread open in case you want to create a separate tracking ticket for this when merging up to the v1.7 branch. Feel free to resolve at your leisure.

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh]: This has been merged and renamed.

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh:22]: Yep, this will produce conflicts anyways.

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh:39]: Since the output was due to the `xtrace` option being enabled (which I want to remove across the board for cleaner logs), I've updated the wording to your last suggestion.

**Comment by alcaeus** [comment]
New patch build: https://spruce.mongodb.com/version/602e562f5623434b3f2cc74c/tasks

Testing against 4.2 and 4.4 passes thanks to some well-placed skips, but I have to investigate the failures on latest as they occur in spec tests. I'd call this done for the time being and revisit the failures once I can see how the 4.2 and 4.4 drivers fare.

**Comment by alcaeus** [comment]
@jmikola Tests now all pass, except for the occasional failure on replica sets due to a double free (our old friend has returned): https://spruce.mongodb.com/version/602f72fc9ccd4e172a60e6e8/tasks

File: tests/FunctionalTestCase.php
Comment (jmikola) [inline tests/FunctionalTestCase.php:54]: Backport? :)

File: .evergreen/run-tests.sh
Comment (jmikola) [inline .evergreen/run-tests.sh:40]: Very nice.

**Review by jmikola — APPROVED** [review]

File: tests/FunctionalTestCase.php
Comment (alcaeus) [reply tests/FunctionalTestCase.php:54]: Yes. Fixed a number of tests 😉

File: .evergreen/run-tests.sh
Comment (alcaeus) [reply .evergreen/run-tests.sh:40]: Yeah, not happy about the formatting, but your suggestion to use groups to skip tests was great.

**Pull request merged**
